File: cpp/src/micro_hashmap_bench.cpp
/// Minimal TaperHashMap microbench — pure insert + probe, no varchar/arena/RowContainer.
/// Tests whether C++ TaperFlatHashTable is inherently slower than Rust TaperHashMap.
///
/// Usage: ./cpp_micro_hashmap [num_keys] [num_probe] [iterations]
/// Default: 100000 keys, 1000000 probes, 10 iterations

#include <cstdio>
#include <cstdint>
#include <cstring>
#include <cstdlib>
#include <chrono>
#include <vector>
#include <random>

#include "taper_hashtable.h"

static constexpr size_t DEFAULT_NUM_KEYS = 100000;
static constexpr size_t DEFAULT_NUM_PROBE = 1000000;
static constexpr size_t DEFAULT_ITERS = 10;

using Clock = std::chrono::high_resolution_clock;
static double elapsed_ms(Clock::time_point t0, Clock::time_point t1) {
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

__attribute__((noinline))
static uint64_t RunInsertProbe(const int64_t* hashes, size_t numRows, size_t numChunks) {
    taper::TaperFlatHashTable table(numChunks);
    uint64_t sum = 0;

    // Use EmplaceBatch with trivial init/update — just count
    table.EmplaceBatch(hashes, static_cast<int32_t>(numRows),
        [](int32_t) { return false; },
        [](uint32_t /*rowIdx*/, char* data) {
            // on_init: write a dummy value (simulates setting row pointer)
            memset(data, 0x42, taper::ROW_PTR_SIZE);
        },
        [&sum](uint32_t /*rowIdx*/, char* data, bool isNew) {
            // on_update: accumulate (simulates agg += value)
            if (!isNew) {
                sum += 1;
            }
        }
    );

    return table.Size() + sum;
}

int main(int argc, char** argv) {
    size_t numKeys = DEFAULT_NUM_KEYS;
    size_t numProbe = DEFAULT_NUM_PROBE;
    size_t numIters = DEFAULT_ITERS;

    if (argc > 1) numKeys = static_cast<size_t>(atoi(argv[1]));
    if (argc > 2) numProbe = static_cast<size_t>(atoi(argv[2]));
    if (argc > 3) numIters = static_cast<size_t>(atoi(argv[3]));

    size_t totalRows = numKeys + numProbe;
    size_t numChunks = 1;
    while (numChunks * 8 < static_cast<size_t>(numKeys / 0.85)) numChunks *= 2;
    // Ensure enough for all distinct keys (keys + some misses from probe)
    while (numChunks * 8 < totalRows) numChunks *= 2;

    fprintf(stderr, "=== C++ Micro HashMap Bench ===\n");
    fprintf(stderr, "numKeys=%zu, numProbe=%zu, totalRows=%zu\n", numKeys, numProbe, totalRows);
    fprintf(stderr, "numChunks=%zu, capacity=%zu\n", numChunks, numChunks * 8);
    fprintf(stderr, "Iterations: %zu (+ 1 warmup)\n\n", numIters);

    // Generate hashes: first numKeys are unique, then numProbe repeat from those
    fprintf(stderr, "Generating hashes...\n");
    std::mt19937_64 rng(42);
    std::vector<int64_t> allHashes(totalRows);
    // Build: unique hashes
    for (size_t i = 0; i < numKeys; i++) {
        allHashes[i] = static_cast<int64_t>(rng());
    }
    // Probe: randomly pick from build hashes (100% hit rate for pure probe test)
    for (size_t i = numKeys; i < totalRows; i++) {
        allHashes[i] = allHashes[rng() % numKeys];
    }
    fprintf(stderr, "Done.\n\n");

    // Warmup
    fprintf(stderr, "Warmup...\n");
    volatile uint64_t w = RunInsertProbe(allHashes.data(), totalRows, numChunks);
    (void)w;
    fprintf(stderr, "Warmup done.\n\n");

    // Timed
    fprintf(stderr, "Running %zu iterations...\n", numIters);
    auto t0 = Clock::now();

    volatile uint64_t checksum = 0;
    for (size_t iter = 0; iter < numIters; iter++) {
        checksum = RunInsertProbe(allHashes.data(), totalRows, numChunks);
    }

    auto t1 = Clock::now();
    double total_ms = elapsed_ms(t0, t1);
    double per_iter_ms = total_ms / numIters;
    double items_per_sec = totalRows / (per_iter_ms / 1000.0);

    printf("=== Results ===\n");
    printf("Total:         %.3f ms\n", total_ms);
    printf("Per iter:      %.3f ms\n", per_iter_ms);
    printf("Items/sec:     %.3f M/s\n", items_per_sec / 1e6);
    printf("Checksum:      %lu\n", (unsigned long)checksum);

    return 0;
}

File: rust/src/bin/micro_hashmap_bench.rs
//! Minimal TaperHashMap microbench — pure insert + probe, no varchar/arena/RowContainer.
//! Tests whether Rust TaperHashMap is inherently faster than C++ TaperFlatHashTable.
//!
//! Usage: ./micro_hashmap_bench [num_keys] [num_probe] [iterations]
//! Default: 100000 keys, 1000000 probes, 10 iterations

use std::time::Instant;
use taper_hashmap::taper_hashmap::TaperHashMap;
use taper_hashmap::chunk::SlotValue;

const DEFAULT_NUM_KEYS: usize = 100000;
const DEFAULT_NUM_PROBE: usize = 1000000;
const DEFAULT_ITERS: usize = 10;

#[inline(never)]
fn run_insert_probe(hashes: &[u64], num_chunks: usize) -> u64 {
    let mut table = TaperHashMap::with_capacity(num_chunks);
    let mut sum: u64 = 0;

    // Use emplace_batch_full with trivial init/update — just count
    table.emplace_batch_full(
        hashes,
        &|_: usize, _: &SlotValue| -> bool { true }, // key_cmp: always match on hash (no real key)
        &mut |_i: usize, sv: &mut SlotValue| {
            // on_init: write dummy value
            sv.bytes = [0x42; 6];
        },
        &mut |_i: usize, _sv: &SlotValue, is_new: bool| {
            // on_update: count hits
            if !is_new {
                sum += 1;
            }
        },
    );

    table.len() as u64 + sum
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let num_keys = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(DEFAULT_NUM_KEYS);
    let num_probe = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(DEFAULT_NUM_PROBE);
    let num_iters = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(DEFAULT_ITERS);

    let total_rows = num_keys + num_probe;
    let mut num_chunks = 1usize;
    while num_chunks * 8 < (num_keys as f64 / 0.85) as usize { num_chunks *= 2; }
    while num_chunks * 8 < total_rows { num_chunks *= 2; }

    eprintln!("=== Rust Micro HashMap Bench ===");
    eprintln!("numKeys={}, numProbe={}, totalRows={}", num_keys, num_probe, total_rows);
    eprintln!("numChunks={}, capacity={}", num_chunks, num_chunks * 8);
    eprintln!("Iterations: {} (+ 1 warmup)\n", num_iters);

    // Generate hashes
    eprintln!("Generating hashes...");
    use rand_mt::Mt19937GenRand64;
    let mut rng = Mt19937GenRand64::new(42);
    let mut all_hashes: Vec<u64> = Vec::with_capacity(total_rows);
    for _ in 0..num_keys {
        all_hashes.push(rng.next_u64());
    }
    for _ in num_keys..total_rows {
        let idx = (rng.next_u64() as usize) % num_keys;
        all_hashes.push(all_hashes[idx]);
    }
    eprintln!("Done.\n");

    // Warmup
    eprintln!("Warmup...");
    let _w = run_insert_probe(&all_hashes, num_chunks);
    eprintln!("Warmup done.\n");

    // Timed
    eprintln!("Running {} iterations...", num_iters);
    let t0 = Instant::now();

    let mut checksum: u64 = 0;
    for _ in 0..num_iters {
        checksum = run_insert_probe(&all_hashes, num_chunks);
    }

    let elapsed = t0.elapsed();
    let total_ms = elapsed.as_secs_f64() * 1000.0;
    let per_iter_ms = total_ms / num_iters as f64;
    let items_per_sec = total_rows as f64 / (per_iter_ms / 1000.0);

    println!("=== Results ===");
    println!("Total:         {:.3} ms", total_ms);
    println!("Per iter:      {:.3} ms", per_iter_ms);
    println!("Items/sec:     {:.3} M/s", items_per_sec / 1e6);
    println!("Checksum:      {}", checksum);
}
